### PR TITLE
Bugfix/firestore functions

### PR DESCRIPTION
**Pull request by baudoin-csp**
This PR introduces main changes for the functions interacting with Firestore.

There are a lot of lines added, removed in the PR in general, and a lot of files modified, added or deleted. In reality, this PR doesn't introduce this amount of code and modifications. A quick analysis shows that it should be around **+1100** lines and **-1000** lines in reality (if we don't count duplicated files, already defined code, and the linter which modify the indentation, pumping these numbers).

Let's break this down:

- The main modifications are in ```firebase/firestore/firestore.ts``` and in ```app/__tests__/firebase.test.tsx```. 
  - Fix the mocking for the ```residence```, ```tenant``` and ```landlord```.
  - Modify the implementations of all functions ```create```, ```get```, ```update``` and ```delete```for: ```TUser```, ```Tenant```, ```Landlord```, ```Residence``` and ```Apartment```. These changes have had the effect of adapting all the related unit tests.
  It concerns **_54 unit tests_** for this part. 
  - The test suite for ```generate_unique_code```might be the more difficult so here is an explanation on what I did:
  I mocked Firestore methods such as `collection`, `addDoc`, `doc`, `getDoc`, and `updateResidence`,
ensuring the tests run in isolation without actual database interactions. One of the main challenges was accurately simulating Firestore’s asynchronous operations and managing the complexity of nested mocks to reflect real data structures. Additionally, spying on `updateResidence` allowed me to verify its invocation while preventing side effects between tests. This process allows `generate_unique_code` to behave correctly under various conditions.
  - The other unit tests are not very hard to understand in my opinion. In case you need more explanations don't hesitate to add a comment 👍 

- Medium to small modifications were made in screens in ```app/screens``` and their unit tests in ```app/__tests__```. These changes mainly regard adapting the usage of firestore functions (signature, usage, etc...).
- Files removed in unit test folder: `CodeApprovedScreen.test.tsx` (deprecated because we don't use this screen anymore), `CodeEntryScreen.test.tsx` (some unit tests are not working), `CreateNewCode.test.tsx` (some unit tests are not working), `cache.test.ts` (duplicated), `firebase.test.tsx` (duplicated).

There is still some modifications that I don't mention in the summary because I consider it is not really important to explain them. I try to keep a concise summary so that you can have a good overview of the work done instead of having 7 long paragraphs to read. If you have any questions don't hesitate :)

### CODE REVIEW

File: app/Navigators/AuthContext.tsx [hunk 1]
```diff
@@ -128,23 +140,19 @@ export function AuthProvider({
...
 export function useAuth() {
   const context = useContext(AuthContext);
   if (context === undefined) {
-    throw new Error('useAuth must be used within an AuthProvider');
+    throw new Error("useAuth must be used within an AuthProvider");
   }
   return context;
-}
\ No newline at end of file
+}
```
Comment (Rcc999) [inline app/Navigators/AuthContext.tsx:158]: Got it! Here's a more positive version:

---

**Comment:**  
Great work on improving the `loadUserData` function! The handling of `tenant` and `landlord` states is clear and well-structured. I also like how you’ve set up the `onSnapshot` listeners to keep everything in sync with Firestore in real-time.

File: app/Navigators/RootNavigator.tsx [hunk 2]
```diff
@@ -9,43 +9,29 @@ import LandlordDrawerNavigator from "./LandlordDrawerNavigator";
...
     }
-}
+  } else {
+    return <AuthStackNavigator />;
+  }
+};
 
-export default RootNavigator;
\ No newline at end of file
+export default RootNavigator;
```
Comment (Rcc999) [inline app/Navigators/RootNavigator.tsx:37]: Nice work on simplifying the RootNavigator logic! The updated flow for determining the navigator based on user.type and the presence of tenant or landlord data is much cleaner and easier to follow. Using initialRouteName for the CodeEntry screen is a great touch to make the navigation intuitive for tenants. This structure looks more robust and should make maintaining the navigation logic simpler

File: app/__tests__/AuthContext.test.tsx
Comment (Rcc999) [inline app/__tests__/AuthContext.test.tsx:1]: Was written by me as part of the S7 sprint, to help baudoin pass the quality gate

File: app/__tests__/CodeEntryScreen.test.tsx
Comment (Rcc999) [inline app/__tests__/CodeEntryScreen.test.tsx:1]: Was written by darius as part of the S7 task, helped baudoin pass the quality gate

File: app/__tests__/ReportScreen.test.tsx
Comment (Rcc999) [inline app/__tests__/ReportScreen.test.tsx:1]: Was written by me as part of the S7 task, helped baudoin pass the quality gate

File: app/__tests__/TenantFormScreen.test.tsx
Comment (Rcc999) [inline app/__tests__/TenantFormScreen.test.tsx:1]: The tests are super detailed and cover all the important scenarios, like form validation, API calls, and error handling. I especially like how you’ve mocked the dependencies properly, which makes the tests really focused and easy to follow. Great job on that!

The requiredFields array is repeated in a few tests, which makes the code a bit longer to read. It’s not a big deal, but combining that into a small helper function could make the tests a bit cleaner and easier to maintain.

File: app/__tests__/cache.test.ts
Comment (Rcc999) [inline app/__tests__/cache.test.ts:1]: The repeated mock setups for FileSystem across multiple tests are fine as they are, but moving them into a shared beforeEach or a helper function could make the test code slightly cleaner and easier to maintain in the future.

File: app/__tests__/firebase.test.tsx
Comment (Rcc999) [inline app/__tests__/firebase.test.tsx:1]: Great Test Coverage: The tests cover so many scenarios, including edge cases and possible errors. It’s clear a lot of thought went into making sure every function is reliable and ready for anything. This kind of detail really shows how much you care about getting things right.

Easy to Follow: The way everything is organized is really nice. Grouping the tests by function makes it super easy to see what's being tested and where. 

Mocks Feel Repetitive: There’s a bit of repetition when it comes to setting up the same mocks (like doc, getDoc, etc.) across multiple tests. It’s not a big deal, but using a shared helper function for these could save some time

In the generate_unique_code function, it’s awesome that tenant codes are being created with addDoc, but there doesn’t seem to be a check to make sure the code is unique before it’s added. Firestore probably handles duplicates, but adding a quick check in the logic could make things clearer and avoid any weird edge cases. Just something to keep in mind for later!

File: app/screens/issues_tenant/ReportScreen.tsx
Comment (Rcc999) [inline app/screens/issues_tenant/ReportScreen.tsx:1]: Nice job keeping the code clean and breaking things into smaller functions! Do you think this structure will make it easy to add new features later?

I noticed some error checks are missing in the tenant ID part. Could this cause issues if something goes wrong?

File: app/screens/landlord/CreateNewCode.tsx
Comment (Rcc999) [inline app/screens/landlord/CreateNewCode.tsx:1]: The check for residence and apartment matching is solid, but the error handling could get repetitive. Could refactoring these checks into a reusable utility function simplify this part of the code?

I really like how you're validating the residence and apartment IDs step by step—it makes the process clear and avoids potential mismatches.

File: firebase/firestore/firestore.ts
Comment (Rcc999) [inline firebase/firestore/firestore.ts:1]: The Firestore functions look really clean and easy to follow. I like how you’ve added clear error messages, which makes it easier to figure out what went wrong if something fails. Do you think adding more comments for the queries would help others understand the logic faster?

The error handling is good, but I noticed that some checks, like for empty query snapshots, are repeated in a few places. Maybe you could create a small helper function to handle those checks and reuse it? What do you think?

**Review by Rcc999 — COMMENTED** [review]
This pull request makes important updates to how Firestore functions are handled. Even though it looks like a lot of changes, most of the added/removed lines come from duplicated files, and formatting.

The main focus is on improving firebase/firestore/firestore.ts and its unit tests. Functions for TUser, Tenant, Landlord, Residence, and Apartment were updated, and 54 tests were adapted. The explanation for the generate_unique_code test is clear and shows a good effort to mock Firestore properly

Some screens and tests were adjusted to match the updated Firestore functions, and unused or duplicate files were cleaned up, which keeps the project more organized.

## Suggestions for Future Improvements:

Test Organization: Consider structuring the test files in folders that mirror the app's folder structure. This will make it easier to locate tests related to specific features or components.

Future Mocking Enhancements: The approach for Firestore mocking is solid, but creating reusable mocks for commonly used functions could save time in future tests.

Overall, this PR does a great job enhancing Firestore functionality and ensuring consistency across the app!

File: app/Navigators/AuthContext.tsx (hunk 1 above)
Comment (baudoin-csp) [reply app/Navigators/AuthContext.tsx:158]: Thank you!

File: app/__tests__/TenantFormScreen.test.tsx
Comment (baudoin-csp) [reply app/__tests__/TenantFormScreen.test.tsx:1]: That's a good point! I'll try this

File: app/__tests__/cache.test.ts
Comment (baudoin-csp) [reply app/__tests__/cache.test.ts:1]: Didn't think about that. Thanks for this suggestion 👍

File: app/__tests__/firebase.test.tsx
Comment (baudoin-csp) [reply app/__tests__/firebase.test.tsx:1]: I tried to setup a shared folder with all the mocks but surprisingly Jest was not working anymore even though the code was almost identical. After some research on forums and documentation I didn't find a good solution so I preferred to let the mock functions inside the unit tests to keep everything independant.

Comment (baudoin-csp) [reply app/__tests__/firebase.test.tsx:1]: That's a good idea! I will implement this

File: app/screens/auth/CodeEntryScreen.tsx
[the remark this answers is not in the document]
Comment (baudoin-csp) [reply app/screens/auth/CodeEntryScreen.tsx:1]: That's a good point! I didn't write the code for this screen. You should either talk about it during our next meeting tomorrow or open an issue.

File: app/screens/issues_landlord/LandlordListIssuesScreen.tsx
[the remark this answers is not in the document]
Comment (baudoin-csp) [reply app/screens/issues_landlord/LandlordListIssuesScreen.tsx:1]: Same comment as for before. Given that I didn't write this screen initially, I think it is more appropriate to open an issue for the original assignee.

File: app/screens/issues_tenant/ReportScreen.tsx
Comment (baudoin-csp) [reply app/screens/issues_tenant/ReportScreen.tsx:1]: I only adapt the code to correctly use the updated firestore functions. The core logic was written by someone else. Consider opening an issue :)

File: app/screens/landlord/CreateNewCode.tsx
Comment (baudoin-csp) [reply app/screens/landlord/CreateNewCode.tsx:1]: Same comment as above

File: firebase/firestore/firestore.ts
Comment (baudoin-csp) [reply firebase/firestore/firestore.ts:1]: After discussion with other teammates, we think that the code is self explanatory. Would you like some comments or is it good for you?

File: app/__tests__/firebase.test.tsx
Comment (Rcc999) [reply app/__tests__/firebase.test.tsx:1]: I see no worries but its a thing to keep in my in the future :)

File: app/screens/auth/CodeEntryScreen.tsx
[the remark this answers is not in the document]
Comment (Rcc999) [reply app/screens/auth/CodeEntryScreen.tsx:1]: Alright no problem

File: firebase/firestore/firestore.ts
Comment (Rcc999) [reply firebase/firestore/firestore.ts:1]: I agree, iIt's more clear for me after discussing it with you!

**Comment by baudoin-csp** [comment]
Thanks @Rcc999 for your review!

Regarding the "test organization", this task is already assigned to @userdarius in this issue [#181](https://github.com/Swent-Fall-2024-team-10/leazy/issues/181). 
and for declaring mocks in a special folder: for the moment it didn't work well and break many tests (more than 30). It is not a big deal for now, if we find a solution I will do another PR to refactor the test.
